src/commands/install-deps.yml

```diff
@@ -26,8 +26,9 @@ steps:
       steps:
         - restore_cache:
             keys:
-              - << parameters.key >>-{{ checksum "./Gemfile.lock"  }}-{{ .Branch }}
-              - << parameters.key >>-{{ checksum "./Gemfile.lock"  }}
+              - << parameters.key >>-{{ .Branch }}-{{ checksum "./Gemfile.lock"  }}
+              - << parameters.key >>-{{ .Branch }}
+              - << parameters.key >>-master
```

Comment:
Let's drop the `master` suffix - both so there's a better catch-all, and to accomadate for users who are opting to use a different primary branch name (e.g `main`).

Reply:
Whatever the primary branch name is we should include it because otherwise it will fall back to the latest snapshot (potentially taken from a random branch). Falling back on the primary ensures that we follow what virtually every PR will do, i.e. forking the primary in a branch.

As a practical solution we can make it into a parameter, named `base-branch`, `primary-branch`, or anything on that lines… I would stick to `master` as the default while it's still the one used by most projects, but of course the final call on that is yours 👍

Reply:
Falling back to whatever the most recent snapshot seems fine. It doesn't matter a whole lot since there's unlikely to be a whole lot of variability. Bundler will make use of whatever is there and clean up the rest. Adding a `base-branch` config seems unnecessary, since if you remove the branch name it shouldn't be necessary to be configurable. It seems like it's adding complexity for little to no gain.

Also, I think it's probably better to have the `checksum` before the `.Branch`. That way we're more likely to get a better cache hit if a branch didn't change the `Gemfile.lock`, or if it matches another branch. For example:

1) Branch B is created off of Branch A
2) Branch B updates the `Gemfile.lock`
3) Branch C is created off of Branch B

It's not a huge deal one way or another, but the best case scenario is that Branch C grabs the cache from Branch B, since it will have the same `Gemfile.lock`. This setup would achieve that:

```yml
- << parameters.key >>-{{ checksum "./Gemfile.lock"  }}-{{ .Branch }}
- << parameters.key >>-{{ checksum "./Gemfile.lock"  }}
- << parameters.key >>
```

Reply:
The consideration about the base-branch is that most projects (both open and closed source) will have the vast majority of topic branches forked from master, targeting the last version of that will likely reap the best results. On the contrary the last snapshot of the generic namespace can refer to a different topic branch, and although I agree that the difference won't be that relevant in most cases, it's still more likely to be a cache-miss.

About the checksum is better to go from more generic to less generic, and it will always be the first build for a given branch, which is more likely to come from the primary branch.

Reply:
> it's still more likely to be a cache-miss.

It's actually almost never going to be a cache miss, but the cache may be incomplete, so bundler will have to make a minor update. This is really only an issue when a branch changes `Gemfile.lock`, and only for the *next* build, so something like:

1) Branch B branches off of `main`
2) Branch B changes `Gemfile.lock`
3) Branch C branches off of `main` and gets a slightly stale cache from Branch B
4) Branch D branches off of `main` and gets a good cache from Branch C

I think in terms of the complexity, the cost of occasionally updating the cache with 1-2 gems is less than the mental overhead of having to manage extra parameters in your configuration. Installing a couple of gems is generally ~10 seconds. Compare that to restoring the cache which is typically 5-10 seconds and often approaches 30 seconds. I don't think the juice is worth the squeeze, especially since it's much less common relative to other branch activity.

> About the checksum is better to go from more generic to less generic,

More generic is likely to be `Gemfile.lock`, though, since that's going to apply across more branches. If you have `.Branch` first then it is more likely to get an incomplete cache, since it's not matching on `Gemfile.lock` from other branches. It'll grab the most recent match, but not the most appropriate one.

Reply:
@mockdeep thanks for the explanation, I think I see where you're coming from and it makes sense 👍 

The way I see the lock-file as more specific is that it's tied to the specific version of many gems, vs. the branch remaining the same across gem changes, and inheriting from the primary branch.

That said it's also more or less bike-shedding at this point, both approaches are good, I just found the one I propose to work better both in a number of closed-source apps and on a virtually the whole @solidusio ecosystem, for which we switched to this kind of approach. I should also remark that in the case of OSS were the lock-file is not checked in the rate at which it changes is way higher (each run will do the equivalent of a `bundle update`).

In the end I don't mind switching to another approach, as I can still fix this locally, and I'll abide to whatever is asked by repo maintainers.

Reply:
> I just found the one I propose to work better both in a number of closed-source apps and on a virtually the whole @solidusio-contrib ecosystem

Can you be more specific about how it is better?

> I should also remark that in the case of OSS were the lock-file is not checked in the rate at which it changes is way higher (each run will do the equivalent of a `bundle update`).

Neither of our approaches will work for cases when there is no lockfile due to the fact that `checksum` throws an error for non-existent files. Falling back to `.Branch` also wouldn't work very well due to the fact that cache keys are immutable, at least until the key expires after 15 days. So any gem updates won't find their way into the cache for `master` until the next expiration:

1) `main` branch builds, no cache exists so it stores it
2) Branch B forks off of `main`, updates a gem, and triggers a new cache of it's own `gemsv1-main`
3) Branch B is merged, triggering a new build on `main`
4) The build on `main` pulls the old cache and bundles, downloading the new dependency. It *does not*, however, push the new dependencies back to the cache, as one already exists for key `gemsv1-main`.

Builds will get progressively more stale caches until the `main` cache expires after 15 days. I think the fix for this is not to cache based on the branch at all, and instead use something like `.Environment.CIRCLE_SHA1`. If we wanted something that is robust for all project types, we could do something like:

```yml
keys:
  - << parameters.key >>-{{ .Branch }}-{{ .Environment.CIRCLE_SHA1 }}
  - << parameters.key >>-{{ .Branch }}
  - << parameters.key >>
```

This will make it so that every build has the best chance of getting a fresh cache. It also solves for #26 without the need for extra configurations.

Reply:
> Can you be more specific about how it is better?

Sure, we sped up builds reducing cache misses considerably.

> Neither of our approaches will work for cases when there is no lockfile due to the fact that checksum throws an error for non-existent files.

That's right, my fault for assuming the lock-file was there. What can be done to consistently have it in all cases is to use `bundle lock`, that generates a lock file without actually installing the gems, that needs to happen before the cache is restored and only takes a few seconds even for large Gemfiles. As an extra benefit it also simplifies the `Bundle Install` step that wouldn't need the check on `Gemfile.lock` anymore. 

I have updated the code to also include an initial `bundle lock` along with the already existing checks on the bundle version, and `bundle clean` after bundle install, that will ensure the cache size won't grow unnecessarily.

I also removed the primary-branch stuff, I still think would be more precise to fall-back on it rather than the generic namespace, but as you said, is also a good compromise with simplicity 👍